snap7/client.py

```diff
@@ -616,6 +616,41 @@ def set_plc_datetime(self, dt):
 
         return self._library.Cli_SetPlcDateTime(self._pointer, byref(buffer))
 
+    def check_as_completion(self, p_value) -> bool:
+        """
+        Method to check Status of an async request
+        :param p_value: Pointer where result of this check shall be written.
+        :return: 0 - Job is done successfully
+        :return: 1 - Job is either pending or contains s7errors
+        """
+        check_result = self._library.Cli_CheckAsCompletion(self._pointer, p_value)
+        try:
+            check_error(check_result, context="client")
+        except Snap7Exception as s7_err:
+            # This error is raised in case of pending job via check_error() method
+            # This error will be accepted/ignored, but others has to fail the test.
+            if check_result == 1 and s7_err.args[0] == b' TCP : Other Socket error (1)':
```

Comment:
From the snap7 library the function looks like this:
```
int S7API Cli_CheckAsCompletion(S7Object Client, int &opResult)
{
    if (Client)
    {
        if (PSnap7Client(Client)->CheckAsCompletion(opResult))
            return JobComplete;
        else
            return JobPending;
    }
    else
		return errLibInvalidObject;
```
If we skip check_error `errLibInvalidObject` can't be caught, as well as `errCliDestroying` in `PSnap7Client(Client)->CheckAsCompletion(opResult)` (haven't found other possible errors).

Reply:
I reviewed the code today, because this `check_error `thing is still annoying me.

I was to much in the Python world and missinterpreted the line with `errCliDestroying`.
The thing is: `JobComplete `== 0, `JobPending `== 1 and `errLibInvalidObject `== -2 for this case.
The `errCliDestroying` thing doesn't have an effect of the execution of this method, because just the errorCode is written in `&opResult`, but a bool is returned.

I guess the repo is just using the Methods from the `snap7_libmain.cpp` file? The only errors which seems to be returned are: `errLibInvalidParam`, which is -1, and `errLibInvalidObject`, which is -2.

That a JobPending is an expected behaviour isn't noticed by error_client and so it does log a pending Job as error.

If I interprete the `Cli_ErrorText` in snap7 correctly, there might be still possible errors been thrown in const longword hex values, which are all greater than 1.

So I guess `check_error(...)` has to be like this:

```
def check_error(code, context="client"):
    if code not in [0, 1]:
        error = error_text(code, context)
        logger.error(error)
        raise Snap7Exception(error)
```

In that case, the ignoreable log.errors() won'tt appear.

Short offtopic: Congrats to your paper @gijzelaerr . ^^

Reply:
hey. Nice find, where did you find the error codes? The `JobPending` status is the missing thing here, and the code you write seems to sound. This also eliminates the need for your code that checks for exceptions. I thought there was something odd there already, but I didn't have the time to dive into it.

PS: regarding the congrats, thanks! (btw it is a PhD ;))

Reply:
`JobPending` and `JobComplete` are in `s7_types.h` defined in snap7.
For the other errors, I followed the `Cli_ErrorText` in snap7 and ended up in `s7_text.cpp` where other errors are listed (but they are partly in different other files, which I won't list here now).

And yes, now this try/except block doesn't make sense any more catching this stuff. 🎉